xcm/src/lib.rs

```diff
@@ -52,6 +53,94 @@ impl Decode for Unsupported {
 	}
 }
 
+/// A single `MultiLocation` value, together with its version code.
+#[derive(Derivative, Encode, Decode)]
+#[derivative(Clone(bound = ""), Eq(bound = ""), PartialEq(bound = ""), Debug(bound = ""))]
+#[codec(encode_bound())]
+#[codec(decode_bound())]
+pub enum VersionedMultiLocation {
+	V0(v0::MultiLocation),
+	V1(v1::MultiLocation),
```

Comment:
Is the intention here to indicate that types like `MultiLocation` can evolve independently from the XCM versioning? If so, then I worry that this may introduce complexity with versioning mismatches, an example being v1 XCM messages working on v2 MultiLocations but not on v0 MultiLocations.

I'm inclined towards a model where all XCM types follow the versioning scheme of XCM in general. That is, if we bump XCM to v2, then all its types would automatically be v2. I realize this is the existing behavior, so perhaps I'm missing some more of the context behind this change?

Reply:
`VersionedMultiLocation` is needed for when we store a `MultiLcation` value between potential runtime upgrades.

Reply:
Ah okay, I see what's going on now. Because these types are exposed in a public API in XCM, there may be situations where an incoming XCM is using an older representation of these types, and we'll need to know how to parse them accordingly.

As for my own concerns, I think I can answer them -- the complexity of maintaining an up-to-date `MultiLocation` or other types is inherent within the design of the system, and the versioning scheme is simply a way to make it explicit and manageable. It's also not going to be as complex as I think it would be, because all incoming XCM messages are always versioned, and each XCM version locks onto a specific version of `MultiLocation` and so is the case for the other types as well.

A follow-up question here is whether we should be seeing more versioned types. I think I only see it being used in `QueryStatus`, but does it make sense for functions that take in a `MultiLocation` to instead take in its versioned counterpart?

Reply:
For dispatchables (i.e. those in transactions) yes it does. Ideally it should be introduced prior to the release of 0.9.10 (XCM v1) since that will change those transaction formats anyway.

As for `QueryStatus`, yes it probably makes sense to introduce versioning to that also.

Reply:
Now introduced in https://github.com/paritytech/polkadot/pull/3693.